### PR TITLE
Initial implementation of honggfuzz.

### PR DESCRIPTION
**Pull request by oliverchang**
Part of #1128.

### CODE REVIEW

File: src/python/bot/fuzzers/honggfuzz/engine.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+import glob
+import os
+import re
+
+from bot.fuzzers import dictionary_manager
+from bot.fuzzers import engine
+from system import environment
+from system import new_process
+
+_DEFAULT_ARGUMENTS = [
```
Comment (inferno-chromium) [inline src/python/bot/fuzzers/honggfuzz/engine.py]: can move this and values 2048, 25 into its own vars in constants.py

File: src/python/bot/fuzzers/honggfuzz/engine.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    '25',
+]
+
+_CRASH_REGEX = re.compile('Crash: saved as \'(.*)\'')
+_HF_SANITIZER_LOG_PREFIX = 'HF.sanitizer.log'
+
+
+def _get_runner():
+  """Get the honggfuzz runner."""
+  honggfuzz_path = os.path.join(environment.get_value('BUILD_DIR'), 'honggfuzz')
```
Comment (inferno-chromium) [inline src/python/bot/fuzzers/honggfuzz/engine.py]: nit: maybe raise an exception if hongfuzz_path does not exist.

File: src/python/bot/fuzzers/honggfuzz/engine.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  return new_process.ProcessRunner(honggfuzz_path)
+
+
+def _find_sanitizer_stacktrace(reproducers_dir):
+  """Find the sanitizer stacktrace from the reproducers dir."""
+  for stacktrace_path in glob.glob(
+      os.path.join(reproducers_dir, _HF_SANITIZER_LOG_PREFIX + '*')):
+    with open(stacktrace_path, 'rb') as f:
+      return f.read()
```
Comment (inferno-chromium) [inline src/python/bot/fuzzers/honggfuzz/engine.py]: Don't we want to append all logs ?

**Comment by inferno-chromium** [comment]
@Dor1s as fyi since he plans AFL migration.

File: src/python/bot/fuzzers/honggfuzz/engine.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+      target_path: Path to the target.
+      build_dir: Path to the build directory.
+
+    Returns:
+      A FuzzOptions object.
+    """
+    arguments = []
+    dict_path = dictionary_manager.get_default_dictionary_path(target_path)
+    if os.path.exists(dict_path):
+      arguments.extend(['--dict', dict_path])
```
Comment (inferno-chromium) [inline src/python/bot/fuzzers/honggfuzz/engine.py]: '--dict' needs to be in constants.py as DICT_FLAG

File: src/python/bot/fuzzers/honggfuzz/engine.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+      max_time: Maximum allowed time for the fuzzing to run.
+
+   Returns:
+      A FuzzResult object.
+    """
+    runner = _get_runner()
+    arguments = _DEFAULT_ARGUMENTS[:]
+    arguments.extend(options.arguments)
+    arguments.extend([
+        '-f',
```
Comment (inferno-chromium) [inline src/python/bot/fuzzers/honggfuzz/engine.py]: nit: does it have longer, more readable option, similar to --run_time ?

File: src/python/bot/fuzzers/honggfuzz/engine.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        '-W',
+        reproducers_dir,
+        '--run_time',
+        str(max_time),
+        '--',
+        target_path,
+    ])
+
+    fuzz_result = runner.run_and_wait(
+        additional_args=arguments, timeout=max_time + 10)
```
Comment (inferno-chromium) [inline src/python/bot/fuzzers/honggfuzz/engine.py]: move 10 to global or constants.py

File: src/python/tests/core/bot/fuzzers/honggfuzz/honggfuzz_engine_test.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        '--run_time',
+        '10',
+        '--',
+        target_path,
+    ], results.command)
+
+    self.assertIn('Seen a crash. Terminating all fuzzing threads', results.logs)
+    self.assertEqual(1, len(results.crashes))
+    crash = results.crashes[0]
+    self.assertEqual(TEMP_DIR, os.path.dirname(crash.input_path))
```
Comment (inferno-chromium) [inline src/python/tests/core/bot/fuzzers/honggfuzz/honggfuzz_engine_test.py]: need to add check to make sure that crash.input_input exists on disk.

File: src/python/bot/fuzzers/honggfuzz/engine.py (hunk 1 above)
Comment (oliverchang) [reply src/python/bot/fuzzers/honggfuzz/engine.py]: I don't think we need a constants.py here :) I think it's better to have vars closer to where they're used for readability, unless they're shared. In this case this file should be quite self contained. Happy to change this if we start needing these vars elsewhere or if you feel strongly.

(moved the timeout and rss limit into their own vars).

File: src/python/bot/fuzzers/honggfuzz/engine.py (hunk 3 above)
Comment (oliverchang) [reply src/python/bot/fuzzers/honggfuzz/engine.py]: There should only be one, so no need to make the code more complex.

File: src/python/bot/fuzzers/honggfuzz/engine.py (hunk 4 above)
Comment (oliverchang) [reply src/python/bot/fuzzers/honggfuzz/engine.py]: Why is dict special compared to all other arguments? :) (This is also the only occurrence).

File: src/python/bot/fuzzers/honggfuzz/engine.py (hunk 5 above)
Comment (oliverchang) [reply src/python/bot/fuzzers/honggfuzz/engine.py]: Yes, done!

File: src/python/bot/fuzzers/honggfuzz/engine.py (hunk 6 above)
Comment (oliverchang) [reply src/python/bot/fuzzers/honggfuzz/engine.py]: Put in a global.

File: src/python/tests/core/bot/fuzzers/honggfuzz/honggfuzz_engine_test.py (hunk 7 above)
Comment (oliverchang) [reply src/python/tests/core/bot/fuzzers/honggfuzz/honggfuzz_engine_test.py]: Done!

File: src/python/bot/fuzzers/honggfuzz/engine.py (hunk 4 above)
Comment (inferno-chromium) [reply src/python/bot/fuzzers/honggfuzz/engine.py]: it is ok, this was just a nit.

File: src/python/bot/fuzzers/honggfuzz/engine.py (hunk 3 above)
Comment (inferno-chromium) [reply src/python/bot/fuzzers/honggfuzz/engine.py]: Ok.

File: src/python/bot/fuzzers/honggfuzz/engine.py (hunk 1 above)
Comment (inferno-chromium) [reply src/python/bot/fuzzers/honggfuzz/engine.py]: Ok, here is fine.

File: src/python/bot/fuzzers/honggfuzz/engine.py (hunk 2 above)
Comment (oliverchang) [reply src/python/bot/fuzzers/honggfuzz/engine.py]: Done!

**Comment by oliverchang** [comment]
/gcbrun. Also added .cc files.

**Pull request merged**